prow/gangway/gangway.go

```diff
@@ -0,0 +1,740 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gangway
+
+import (
+	context "context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	codes "google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	status "google.golang.org/grpc/status"
+	coreapi "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	prowcrd "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pjutil"
+	"k8s.io/test-infra/prow/version"
+)
+
+const (
+	HEADER_API_CONSUMER_TYPE = "x-endpoint-api-consumer-type"
+	HEADER_API_CONSUMER_ID   = "x-endpoint-api-consumer-number"
+)
+
+type Gangway struct {
+	UnimplementedProwServer
+	ConfigAgent              *config.Agent
+	ProwJobClient            ProwJobClient
+	InRepoConfigCacheHandler *config.InRepoConfigCacheHandler
+}
+
+// ProwJobClient is mostly for testing (for calling into the low-level
+// Kubernetes API to check whether gangway behaved correctly).
+type ProwJobClient interface {
+	Create(context.Context, *prowcrd.ProwJob, metav1.CreateOptions) (*prowcrd.ProwJob, error)
+}
+
+func (gw *Gangway) CreateJobExecution(ctx context.Context, req *CreateJobExecutionRequest) (*JobExecution, error) {
+	err, md := getHttpRequestHeaders(ctx)
+	if err != nil {
+		logrus.WithError(err).Error("could not find request HTTP headers")
```

Comment:
I think the original reason why I used Error level instead of Debug is because I was expecting to only deploy Gangway behind a ESPv2 proxy (which would on its own also handle header validation, etc). So in that scenario, if we get the wrong headers or some headers are missing, it would imply that ESPv2 is misbehaving, which is something we would want to know about from Prow's perspective.

That being said not everyone will use a ESPv2 proxy, and for those cases I suppose it makes more sense to lower the log level to Info or Debug because we don't care that much about human error (where our code is WAI). Was this the rationale behind your desire to lower the log level?

Reply:
yes lower the log level is preferred, simply put: user errors are not supposed to catch operator's attention every time